### ISSUE TITLE
Backdrop doesn't consider version 2 of a module to be an update of the latest version 1.

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#5563]
This problem came up after the rules + entity_token debacle. 
See https://github.com/backdrop-contrib/rules/issues/178 
(or maybe someone can write up a better description)

**Comment by BWPanda** [comment]
Doesn't an issue need a near-ready PR to go into the next bug-fix milestone...?

Also, related: https://github.com/backdrop/backdrop-issues/issues/5562

**Comment by argiepiano** [comment]
Related to my question #5562. To reproduce:

1. Install an "old" major version of a contrib module. For example, [Entity Tokens version 1.1.0-beta](https://github.com/backdrop-contrib/entity_token/releases/tag/1.x-1.1.0-beta)
2. (The current supported version of Entity Tokens is version 2.0.1)
3. Visit admin/reports/updates. The old major version is listed as "Up to date", and version 2.0.1 is listed as "Also available".

An issue happens when you have Entity Tokens version 1 installed, and you have another, outdated module  such as Rules, that depends on Entity Tokens.  The latest version of Rules  depends on >= Entity Tokens 2.0.0. When you try to update Rules,  the update fails, since Entity Tokens version 1 is considered "Up to date" and the update process doesn't try to update Entity Tokens to version 2.

Additionally, you can't use Project Installer to install the next major version of a module. You are forced to uninstall and manually remove the old major version from the modules folder, and then install the next major version.

With Drupal 7, if the version was unsupported, you could use the installer to upgrade.

**Comment by jenlampton** [comment]
> Doesn't an issue need a near-ready PR to go into the next bug-fix milestone...?

You're right. I just consulted the documentation, and it doesn't sate anything about critical bug fixes!  I'm going to remove the milestone for now, but after the documentation gets updated we can add it back.

**Comment by argiepiano** [comment]
After a bit of digging inside the update module, I don't think this is a bug.  The idea is that a module could have two working major versions, and people could choose among versions (e.g. Commerce Kickstart - while not a module, but rather a profile - has both versions 1 and 2 as possible working versions).  

However, I think there is a way for the team managing the project portal at Backdrop to mark a version as "unsupported", which will in theory show the next major version as a required update. Can someone with access to the module portal confirm this? And if so, can you mark Entity Tokens version 1 as unsupported?

**Comment by indigoxela** [comment]
Loosely related: https://github.com/backdrop-ops/backdropcms.org/issues/474

**Comment by bugfolder** [comment]
There are settings for the project to both specify which major version is "Recommended" and also whether any given release is "Supported." Here's the page for Entity Token:

![supported](https://user-images.githubusercontent.com/29214084/159714937-2096cf08-8b4d-4d64-a2db-430aac301a06.jpg)

I will uncheck "Supported" for version 1. Does that give the desired behavior? I see that it is still offered as "other releases" on https://backdropcms.org/project/entity_token.

**Comment by argiepiano** [comment]
@bugfolder thanks for unchecking "unsupported" for Entity Tokens. I'm sad to report that this had made no difference in the status of the module in my test site (see attached). I've tried clearing caches and running cron. 

The code in the upgrade module is supposed to account for unsupported modules and output something to indicate that. So there may actually be a bug there. 
<img width="1190" alt="Screen Shot 2022-03-23 at 9 37 41 AM" src="https://user-images.githubusercontent.com/9938978/159738160-15ac9639-d384-413e-97c7-b58aedbc1aa9.png">

**Comment by argiepiano** [comment]
@jenlampton, given that this was not a bug, and that your issue is now solved by @bugfolder setting version 1 of Entity Tokens as unsupported, I believe we can close this bug report. Please feel free to reopen if I'm missing something here.

**Comment by bugfolder** [comment]
I'll leave this issue closed, but it seems like there's two cases where b.org administrators need to step in to modify a project node directly, and module maintainers need to know they have to ask administrators to take action:

* When the major release version changes (e.g., from 1.x.x to 2.0.0);
* When an older major version needs to be marked as unsupported.

So I think some documentation is called for.  This seems like the right page for it:

https://docs.backdropcms.org/documentation/contribute-to-a-module-layout-or-theme

**Comment by bugfolder** [comment]
New text added to that page. If anyone sees anything amiss, please create an issue in the docs.backdropcms.org issue queue (or if it's small or obvious and you can, just fix it directly). Here's a copy of the new text (at the bottom of the section).

![new text](https://user-images.githubusercontent.com/29214084/159759686-d25dbc5c-b896-4d7a-acbf-1e00929b8962.jpg)

**Comment by jenlampton** [comment]
This is definitely still a bug that needs to be addressed. Though it's supported to have both supported and recommended versions of modules, the project browser shouldn't fail quite so terribly for these kinds of updates. I'd like to think about how we might be able to provide the option to switch to different major version for one of their modules.

**Comment by argiepiano** [comment]
@jenlampton  once a major release is marked as unsupported, the installer will allow you to install the supported major version.  In the case that you cited in your OP, updating Rules will also update Entity Tokens to version 2, now that version 1 is unsupported.

**Comment by jenlampton** [comment]
Because the problem remains for modules that have two (or more) supported major versions, we need to make the user-experience better for people in those scenarios.

**Comment by klonos** [comment]
Life/work is crazy ATM, but taking the time to quickly go through some issues in the core queue, and this one caught my eye. Just wanted to quickly mention that Drupal now supports a `lifecycle` entry in .info.yml files (change record: [Info files can now contain 'lifecycle' and 'lifecycle_link' keys to convey the stability of a module/theme](https://www.drupal.org/node/3215042)). It seems that something like `lifecycle = deprecated` is what we'd need in order to mimic the same functionality:

> - **deprecated**
>   Something on the way out. You can still install it, but there are warnings if it's installed on your site.

...as the title of the Drupal change record mentions, there's also a `lifecycle_link = https://link.to.change/record` property, which sounds cool/useful and could be exposed in the admin UI.

Related:
- #5367
- https://github.com/backdrop-ops/backdropcms.org/issues/100
